crates/uv-resolver/src/resolution.rs

```diff
@@ -186,79 +213,136 @@ impl ResolutionGraph {
                 PubGrubPackage::Package(package_name, None, Some(url)) => {
                     // Create the distribution.
-                    let pinned_package = if let Some((editable, _, _)) = editables.get(package_name)
-                    {
+                    let dist = if let Some((editable, _, _)) = editables.get(package_name) {
                         Dist::from_editable(package_name.clone(), editable.clone())?
                     } else {
                         Dist::from_url(package_name.clone(), url_to_precise(url.clone()))?
                     };
 
-                    // Add its hashes to the index, preserving those that were already present in
-                    // the lockfile if necessary.
-                    if let Some(digests) = preferences
+                    // Extract the hashes, preserving those that were already present in the
+                    // lockfile if necessary.
+                    let hashes = if let Some(digests) = preferences
                         .match_hashes(package_name, version)
                         .filter(|digests| !digests.is_empty())
                     {
-                        hashes.insert(package_name.clone(), digests.to_vec());
-                    } else if let Some(metadata_response) =
-                        distributions.get(&pinned_package.version_id())
-                    {
+                        digests.to_vec()
+                    } else if let Some(metadata_response) = distributions.get(&dist.version_id()) {
                         if let MetadataResponse::Found(ref archive) = *metadata_response {
                             let mut digests = archive.hashes.clone();
                             digests.sort_unstable();
-                            hashes.insert(package_name.clone(), digests);
+                            digests
+                        } else {
+                            vec![]
                         }
-                    }
+                    } else {
+                        vec![]
+                    };
+
+                    // Extract the metadata.
+                    let metadata = if let Some((_, metadata, _)) = editables.get(package_name) {
+                        metadata.clone()
+                    } else {
+                        let dist = PubGrubDistribution::from_url(package_name, url);
+
+                        let response = distributions.get(&dist.version_id()).unwrap_or_else(|| {
+                            panic!(
+                                "Every package should have metadata: {:?}",
+                                dist.version_id()
+                            )
+                        });
+
+                        let MetadataResponse::Found(archive) = &*response else {
+                            panic!(
+                                "Every package should have metadata: {:?}",
+                                dist.version_id()
+                            )
+                        };
```

Comment:
Not necessarily asking for any changes here, but just a "shout out loud": I feel like I've seen this pattern a lot where the code for processing distributions repeats itself. This looks almost identical to the metadata extraction above for example. (I don't yet have opinions on how to resolve this or even if it's worth resolving.)

Reply:
Yeah it's nearly identical. I'm gonna look at DRYing it up.

Reply:
Sorry, to be clear: it's exactly the same. They're different arms in a match.